### PR TITLE
Fixes ILU0 when using flexible solvers.

### LINKED ISSUE #2490: FlexibleSolver with defaults never recalculates ILU pattern.
We end up with
```
Dune::DummyUpdatePreconditioner<Opm::ParallelOverlappingILU0<Dune::BCRSMatrix<Opm::MatrixBlock<double, 3, 3>, std::allocator<Opm::MatrixBlock<double, 3, 3> > >, Dune::BlockVector<Dune::FieldVector<double, 3>, std::allocator<Dune::FieldVector<double, 3> > >, Dune::BlockVector<Dune::FieldVector<double, 3>, std::allocator<Dune::FieldVector<double, 3> > >, Dune::Amg::SequentialInformation> >::update
```
which does nothing and we run into convergence issues for the simplest cases.

### PR DESCRIPTION
**Pull request by blattms**
when running flow_blackoil_dunecpr with the defaults we experienced severs convergence issues with SPE1CASE2, because the decomposition was never recalculated. This is fixed in PR.

Closes #2490

### CODE REVIEW

**Comment by blattms** [comment]
Not sure whether this is even built by jenkins.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
Needs some more testing before merge.

**Comment by blattms** [comment]
Locally tested. Convergence issues are gone with these changes.  Defaults are still not the same (e.g. --ilu-relaxation=1.0 versus flow using --ilu-relaxation=0.9). Ready for merge,

File: opm/simulators/linalg/PreconditionerFactory.hpp [hunk 1]
```diff
@@ -156,26 +156,20 @@ class PreconditionerFactory
         using C = Comm;
         doAddCreator("ILU0", [](const O& op, const P& prm, const C& comm) {
             const double w = prm.get<double>("relaxation");
-#if DUNE_VERSION_NEWER(DUNE_ISTL, 2, 7)
-            return wrapBlockPreconditioner<DummyUpdatePreconditioner<SeqILU<M, V, V>>>(comm, op.getmat(), w);
-#else
-            return wrapBlockPreconditioner<DummyUpdatePreconditioner<SeqILU0<M, V, V>>>(comm, op.getmat(), w);
-#endif
+            return std::make_shared<Opm::ParallelOverlappingILU0<M, V, V, C>>(
```
Comment (atgeirr) [inline opm/simulators/linalg/PreconditionerFactory.hpp]: This change (together with the "ILUn" change below) seems to eliminate the option to run the Block Jacobi/SeqILU preconditioner combination?

Comment (blattms) [reply opm/simulators/linalg/PreconditionerFactory.hpp]: Well, flow does not use that combination either (see ISTLSolverEbos.hpp). If I am not completely mistaking `ParallelOverlappingILU0` should do the same if the matrix is correctly set up. My aim is to bring the codes paths closer together as they have been different  from the beginning and are diverging more and more...

The previous version using `DummyUpdatePreconditioner` never recalculated the decomposition as its update method is a no-op.

Comment (atgeirr) [reply opm/simulators/linalg/PreconditionerFactory.hpp]: > Well, flow does not use that combination either (see ISTLSolverEbos.hpp). If I am not completely mistaking `ParallelOverlappingILU0` should do the same if the matrix is correctly set up. My aim is to bring the codes paths closer together as they have been different from the beginning and are diverging more and more...

I know regular Flow does not use that combination, just thought it was nice to offer the option (more options to experiment with is better). Which code paths is it you want to bring together?

**Comment by blattms** [comment]
As it was, it was basically broken (admittedly, I did not check the parallel version).

The code paths are the various ISTLSolver*.hpp files which have diverged quite dramatically and will otherwise continue to do so.

**Comment by blattms** [comment]
jenkins build this serial please

**Comment by blattms** [comment]
If you insist I will add the ILU0 and ILUn stuff again...

**Comment by atgeirr** [comment]
> If you insist I will add the ILU0 and ILUn stuff again...

No I will not insist. I'll merge now. If we decide we want to re-add it, it would not be hard. It may be a little bit misleading to have both "ILU0" and "ParOverILU0" yield the same preconditioner, but it's not a problem, and it makes the change backwards compatible.

**Comment by blattms** [comment]
Cool, thanks.

**Pull request merged**